File: api/src/spp_extras_api/utils/loremaster.py
import json
from from_root import from_root
with open(from_root('data/achievements/loremasterAchCriteria.json'), 'r') as json_file:
    lm_criteria = json.load(json_file)
with open(from_root('data/zones/zoneContinents.json'), 'r') as json_file:
    zone_continents = json.load(json_file)


# See which completed quests belong to which criteria and what their counts are
def loremaster(completed_quests, template_quests, loremaster_prog):
    alliance_template_quests = template_quests['alliance']
    horde_template_quests = template_quests['horde']
    neutral_quests = template_quests['neutral']
    all_alliance_template_quests = {
        **alliance_template_quests, **neutral_quests}
    all_horde_template_quests = {
        **horde_template_quests, **neutral_quests}

    # Track highest counts and most recent dates for main Loremaster quests
    # Use counts from misc Loremaster criteria as a starting point
    alliance_ek = loremaster_prog['1676']
    alliance_ek_count = alliance_ek['count']
    alliance_ek_date = alliance_ek['date']
    alliance_k = loremaster_prog['1678']
    alliance_k_count = alliance_k['count']
    alliance_k_date = alliance_k['date']
    horde_ek = loremaster_prog['1677']
    horde_ek_count = horde_ek['count']
    horde_ek_date = horde_ek['date']
    horde_k = loremaster_prog['1680']
    horde_k_count = horde_k['count']
    horde_k_date = horde_k['date']

    # Track criteria counts and dates for each faction/zone
    all_criteria = {'alliance': {}, 'horde': {}}

    def count_sub_criteria(criteria_id, date, faction):
        if criteria_id not in all_criteria[faction]:
            all_criteria[faction][criteria_id] = {
                'count': 1,
                'date': date
            }
        else:
            all_criteria[faction][criteria_id]['count'] += 1
            existing_date = all_criteria[faction][criteria_id]['date']
            if existing_date < date:
                all_criteria[faction][criteria_id]['date'] = date

    for quest_id in completed_quests:
        date = completed_quests[quest_id]['timer']

        # Count Alliance progress
        if quest_id in all_alliance_template_quests:
            template_quest = all_alliance_template_quests[quest_id]
            zone_id = str(template_quest['zoneorsort'])
            if zone_id in lm_criteria['alliance']:
                zone_criteria = lm_criteria['alliance'][zone_id]
                criteria_id = str(zone_criteria['criteria'])
                ach_id = zone_criteria['achievement']
                count_sub_criteria(criteria_id, date, 'alliance')

                # Add to main counters
                if ach_id == 1676:
                    alliance_ek_count += 1
                    if alliance_ek_date < date:
                        alliance_ek_date = date
                elif ach_id == 1678:
                    alliance_k_count += 1
                    if alliance_k_date < date:
                        alliance_k_date = date

        # Count Horde progress
        if quest_id in all_horde_template_quests:
            template_quest = all_horde_template_quests[quest_id]
            zone_id = str(template_quest['zoneorsort'])
            if zone_id in lm_criteria['horde']:
                zone_criteria = lm_criteria['horde'][zone_id]
                criteria_id = str(zone_criteria['criteria'])
                ach_id = zone_criteria['achievement']
                count_sub_criteria(criteria_id, date, 'horde')

                # Add to main counters
                if ach_id == 1677:
                    horde_ek_count += 1
                    if horde_ek_date < date:
                        horde_ek_date = date
                elif ach_id == 1680:
                    horde_k_count += 1
                    if horde_k_date < date:
                        horde_k_date = date

    new_loremaster_prog = {
        '1676': {'count': alliance_ek_count, 'date': alliance_ek_date},
        '1678': {'count': alliance_k_count, 'date': alliance_k_date},
        '1677': {'count': horde_ek_count, 'date': horde_ek_date},
        '1680': {'count': horde_k_count, 'date': horde_k_date}
    }

    all_progress = {
        'main_prog': new_loremaster_prog,
        'sub_prog': all_criteria
    }

    return all_progress


# Some criteria for Loremaster achievements don't have matching zonerefs or have duplicates
# Progress for these criteria are tracked separately from the other Loremaster criteria
# and are tracked like shared achievements
# There will likely be side-effects to achievements tracked this way since there's no
# way to check for duplicate progress, but the alternative is risking achievement credit
# not being given on transfer at all
def misc_lm_criteria(criteria_id):
    alliance_ek = [5903, 5910, 9398, 9424, 5927, 5928, 5940, 5944, 9422]
    alliance_k = [7895, 6017, 6021, 6030]
    horde_ek = [5955, 5962, 9425, 5979, 5980, 5992, 5996, 9423]
    horde_k = [7899, 6112, 6116, 6125]
    
    if criteria_id in alliance_ek:
        return '1676'
    elif criteria_id in alliance_k:
        return '1678'
    elif criteria_id in horde_ek:
        return '1677'
    elif criteria_id in horde_k:
        return '1680'


# See if Loremaster achievement is earned after sharing progress
def loremaster_earned(ach_id, count):
    # Alliance Eastern Kingdoms
    if ach_id == 1676 and count >= 700:
        return True
    # Alliance Kalimdor
    elif ach_id == 1678 and count >= 700:
        return True
    # Horde Eastern Kingdoms
    elif ach_id == 1677 and count >= 550:
        return True
    # Horde Kalimdor
    elif ach_id == 1680 and count >= 685:
        return True
    else:
        return False

File: src/utils/loremaster.ts
import _lmCriteria from '../../data/achievements/loremasterAchCriteria.json';
import {
  AchProgress,
  AllTemplateQuests,
  CompletedRegQuests,
  Faction,
  LMCriteriaTracker,
  LoremasterCriteria,
} from '../types';

const lmCriteria = _lmCriteria as LoremasterCriteria;

// See which completed quests belong to which criteria and what their counts are
export const loremaster = (
  completedQuests: CompletedRegQuests,
  templateQuests: AllTemplateQuests,
  loremasterProg: AchProgress,
) => {
  const allianceTemplQuests = templateQuests.alliance;
  const hordeTemplQuests = templateQuests.horde;
  const neutralTemplQuests = templateQuests.neutral;
  const allAllianceTemplQuests = { ...allianceTemplQuests, ...neutralTemplQuests };
  const allHordeTemplQuests = { ...hordeTemplQuests, ...neutralTemplQuests };

  // Track highest counts and most recent dates for main Loremaster quests
  // Use counts from misc Loremaster criteria as a starting point
  const allianceEK = loremasterProg['1676'];
  let allianceEKCount = allianceEK.counter;
  let allianceEKDate = allianceEK.date;
  const allianceK = loremasterProg['1678'];
  let allianceKCount = allianceK.counter;
  let allianceKDate = allianceK.date;
  const hordeEK = loremasterProg['1677'];
  let hordeEKCount = hordeEK.counter;
  let hordeEKDate = hordeEK.date;
  const hordeK = loremasterProg['1680'];
  let hordeKCount = hordeK.counter;
  let hordeKDate = hordeK.date;

  // Track criteria counts and dates for each faction/zone
  const allCriteria: LMCriteriaTracker = { alliance: {}, horde: {} };

  const countSubCriteria = (criteriaID: number, date: number, faction: Faction) => {
    if (!allCriteria[faction][criteriaID]) {
      allCriteria[faction][criteriaID] = { counter: 1, date: date };
    } else {
      allCriteria[faction][criteriaID]['counter']++;
      const existingDate = allCriteria[faction][criteriaID]['date'];
      if (existingDate < date) allCriteria[faction][criteriaID]['date'] = date;
    }
  }

  for (let questID in completedQuests) {
    const date = completedQuests[questID]['timer'] || 0;

    // Count Alliance progress
    const allianceTemplQuest = allAllianceTemplQuests[questID];
    if (allianceTemplQuest) {
      const zoneID = allianceTemplQuest.ZoneOrSort;
      const zoneCriteria = lmCriteria['alliance'][zoneID];
      const criteriaID = zoneCriteria.criteria;
      const achID = zoneCriteria.achievement;
      countSubCriteria(criteriaID, date, 'alliance');

      // Add to main counters
      if (achID === 1676) {
        allianceEKCount++;
        if (allianceEKDate < date) allianceEKDate = date
      } else if (achID === 1678) {
        allianceKCount++;
        if (allianceKDate < date) allianceKDate = date;
      }
    }

    // Count Horde progress
    const hordeTemplQuest = allHordeTemplQuests[questID];
    if (hordeTemplQuest) {
      const zoneID = hordeTemplQuest.ZoneOrSort;
      const zoneCriteria = lmCriteria['horde'][zoneID];
      const criteriaID = zoneCriteria.criteria;
      const achID = zoneCriteria.achievement;
      countSubCriteria(criteriaID, date, 'horde');

      // Add to main counters
      if (achID === 1677) {
        hordeEKCount++;
        if (hordeEKDate < date) hordeEKDate = date;
      } else if (achID === 1680) {
        hordeKCount++;
        if (hordeKDate < date) hordeKDate = date;
      }
    }
  }

  const newLoremasterProg = {
    1676: { count: allianceEKCount, date: allianceEKDate },
    1678: { count: allianceKCount, date: allianceKDate },
    1677: { count: hordeEKCount, date: hordeEKDate },
    1680: { count: hordeKCount, date: hordeKDate },
  };

  const allProgress = {
    mainProg: newLoremasterProg,
    subProg: allCriteria,
  };

  return allProgress;
}

// Some criteria for Loremaster achieves don't have matching zonerefs or have duplicates.
// Progress for these criteria are tracked separately from the other Loremaster criteria
// and are tracked like shared achievements.
// There will likely be side-effects to achievements tracked this way since there's no
// way to check for duplicate progress, but the alternative is risking achievement credit
// not being given on transfer at all.
export const miscLMCriteria = (criteriaID: number) => {
  const allianceEK = [5903, 5910, 9398, 9424, 5927, 5928, 5940, 5944, 9422];
  const allianceK = [7895, 6017, 6021, 6030];
  const hordeEK = [5955, 5962, 9425, 5979, 5980, 5992, 5996, 9423];
  const hordeK = [7899, 6112, 6116, 6125];
  if (allianceEK[criteriaID]) {
    return 1676;
  } else if (allianceK[criteriaID]) {
    return 1678;
  } else if (hordeEK[criteriaID]) {
    return 1677;
  } else if (hordeK[criteriaID]) {
    return 1680;
  }
}

// See if Loremaster achievement is earned after sharing progress
export const loremasterEarned = (achID: number, count: number) => {
  // Alliance Eastern Kingdoms
  if (achID === 1676 && count >= 700) return true;
  // Alliance Kalimdor
  else if (achID === 1678 && count >= 700) return true;
  // Horde Eastern Kingdoms
  else if (achID === 1677 && count >= 550) return true;
  // Horde Kalimdor
  else if (achID === 1680 && count >= 685) return true;
  else return false;
}
